### PR TITLE
Add mypy typings

### PR DESCRIPTION
**Pull request by libre-man**
The issuer is not enough to accurately find a registration. For example, each
hosted Canvas instance has the same issuer.

### CODE REVIEW

**Comment by libre-man** [comment]
By the way, I also have created some typing stubs for the repository, that I would be more than happy to move in to the actual code. This way you could also add `mypy` to your CI setup, however supporting python 2 would be a bit more difficult.

**Comment by dmitry-viskov** [comment]
hi @libre-man.

First of all - thank you for your pull request!

The same thoughts was in my head in the beginning of development this library. There are few concerns:

1. As i understand this default value could be changed through config: https://github.com/instructure/canvas-lms/blob/master/config/security.yml.example#L4 . Am I wrong?

2. By default on the `login` step we don't know about the client-id in common case. For example if you check this library with lti advantage certification suite you will see that client-id is not passed:

login

I understand that canvas pass this value - but it looks like exception case.

3.  >This way you could also add `mypy` to your CI setup, however supporting python 2 would be a bit more difficult.

I would like to support Python 2 because unfortunately [edX](https://github.com/edx/edx-platform) still works on python 2 and one of the reasons of creation this library was ability to configure edX as LTI 1.3 Tool

Anyway we may expand:

```def find_registration_by_issuer(self, iss):```

with new param:

```def find_registration_by_issuer(self, iss, request_params=None): ```

It will be `GET`/`POST` params in case of `login` and JWT data in case of `launch`.
And after this you may create your custom `ToolConf` that will implement custom logic using `request_params`. What do you think?

**Comment by libre-man** [comment]
Thanks for your extensive answer!

1. Yes you could change it that way, however instructure does not change this for their production envs: https://canvas.instructure.com/doc/api/file.lti_dev_key_config.html#step-2. So every Canvas customer has the same `iss`.

2. That is the unfortunate case, however the [spec](https://www.imsglobal.org/spec/lti/v1p3/#client_id-login-parameter) does list `client_id` as a optional parameter for the login step.

3. We could also add the typings as type comments, that way you could still support python 2, however that will take me a bit longer unfortunately, as I have never used mypy that way before.

-------

I would prefer your option of adding a optional parameter to `find_registration_by_issuer`, however that will break all the code that now implements the abstract class, as the existing implementations of `find_registration_by_issuer` don't have this parameter. I do really like the idea of passing all information to the config class, as that way one could use all available information.

**Comment by dmitry-viskov** [comment]
@libre-man thanks for adding type-checking! i'll review all changes a look a little bit later.

> however that will break all the code that now implements the abstract class, as the existing implementations of find_registration_by_issuer don't have this parameter.

i don't think that it will break anything. We just add new optional argument in 2 places:

```
class ToolConfAbstract(object):
    ...
    @abstractmethod
    def find_registration_by_issuer(self, iss, params=None):
        raise NotImplementedError

class ToolConfDict(ToolConfAbstract):
    ...
    def find_registration_by_issuer(self, iss, params=None):
        ...
```

And in the `oidc_login.py` / `message_launch.py` we add:

```
try:
    registration = self._tool_config.find_registration_by_issuer(iss, params)
except TypeError:
    registration = self._tool_config.find_registration_by_issuer(iss)  # backward compatibility
```

Then in you project you will implement your own `CanvasConfToolConf(ToolConfAbstract)` that will operate with `params` to find registration

**Comment by libre-man** [comment]
That could be a solution, however that will run the `find_registration_by_issuer` function twice if someone's implementation of that function throws a `TypeError` on certain conditions. I think it would be best, if you would take this route, to add a `**kwargs`` param to the `find_registration_by_issuer` method to prevent such problems in the future.

File: pylti1p3/message_launch.py [hunk 1]
```diff
@@ -313,22 +457,29 @@ def validate_nonce(self):
...
+        # type: (T_SELF) -> T_SELF
         iss = self._get_iss()
 
+        # Check client id
+        aud = self._get_jwt_body().get('aud')
+        client_id = aud[0] if isinstance(aud, list) else aud
+
         # Find registration
-        self._registration = self._tool_config.find_registration_by_issuer(iss)
+        self._registration = self._tool_config.find_registration_by_issuer_and_client_id(iss, client_id)
```
Comment (dmitry-viskov) [inline pylti1p3/message_launch.py]: As we discussed before - let's reuse `find_registration_by_issuer` with additional param

File: pylti1p3/oidc_login.py [hunk 2]
```diff
@@ -104,8 +137,13 @@ def validate_oidc_login(self):
...
             raise OIDCException('Could not find login hint')
 
+        # Many LMS systems send this to more uniquely identify the
+        # registration, as for example Canvas uses the same `iss` for all
+        # hosted instances.
+        client_id = self._get_request_param('client_id')
+
         # fetch registration details
-        registration = self._tool_config.find_registration_by_issuer(iss)
+        registration = self._tool_config.find_registration_by_issuer_and_client_id(iss, client_id)
```
Comment (dmitry-viskov) [inline pylti1p3/oidc_login.py]: The same here - let's reuse `find_registration_by_issuer`

File: pylti1p3/grade.py [hunk 3]
```diff
@@ -59,4 +88,7 @@ def get_value(self):
             'timestamp': self._timestamp,
             'userId': self._user_id
         }
-        return json.dumps({k: v for k, v in data.items() if v})
+        if self._extra_claims is not None:
+            data.update(self._extra_claims)
+
+        return json.dumps({k: v for k, v in data.items() if v is not None})
```
Comment (dmitry-viskov) [inline pylti1p3/grade.py]: It was my mistake. Thanks for fixing! 👍

File: pylti1p3/deep_link.py [hunk 4]
```diff
@@ -29,19 +60,28 @@ def get_message_jwt(self, resources):
...
+        assert private_key is not None, 'Registration private key should not be None'
+        return jwt.encode(message, private_key, algorithm='RS256')
 
     def get_response_jwt(self, resources):
+        # type: (t.Sequence[DeepLinkResource]) -> str
         message_jwt = self.get_message_jwt(resources)
-        return self.encode_jwt(message_jwt)
+        encoded_jwt = self.encode_jwt(message_jwt)
+        if sys.version_info[0] > 2:
+            return encoded_jwt.decode('utf-8')
```
[the remark this answers is not in the document]
Comment (libre-man) [reply pylti1p3/deep_link.py]: That seems like a good plan.

File: MANIFEST.in [hunk 5]
```diff
@@ -0,0 +1,2 @@
+include pylti1p3/py.typed
+include pylti1p3/contrib/py.typed
```
[the remark this answers is not in the document]
Comment (libre-man) [reply MANIFEST.in]: Yes, sounds like a plan!

**Comment by libre-man** [comment]
I've changed the tool config as you requested, however there are a few problems/differences:
- I've made it possible for users to disable the fallback, to make it possible to disable any exception swallowing that would occur in code that does implement the new parameter.
- I've made the extra parameter a function, as it seems impossible for the `OIDCLogin` class to get all request parameters.
- Using this function will be quite annoying I think, as the data available is completely dependent on in which context the config provider is used. I will be implementing two config classes, one for the login, and one for the actual launch. I don't know if there is anyway around that though, without splitting the method.

**Comment by dmitry-viskov** [comment]
> I've changed the tool config as you requested, however there are a few problems/differences:
> 

Please check my solution: https://github.com/dmitry-viskov/pylti1.3/commit/20c8d8d56ab0b5f63d7ad5487b067609d85cb83a
* backward compatibility is supported
* instead of pass function into `find_registration` method let's pass `action` argument. Depending on this you may know what params are passed in the `kwargs`

**Comment by dmitry-viskov** [comment]
hi @libre-man.

Let's focus here on integration `typing` and `mypy`. You did a lot of work and it will be good to merge this big feature into the main branch. 
Could you please make rebase, fix conflicts and remove custom logic related to the `find_registration_by_issuer`. I've made fix: https://github.com/dmitry-viskov/pylti1.3/commit/b3b2783fb1a92cb7a72262b5155a5039b40c2bb6 - so you may define your own `def find_registration_by_issuer(self, iss, action, request, jwt_body, *args, **kwargs)` method in `ToolConf` and retrieve `client_id` depending on this data

**Comment by libre-man** [comment]
Hi @dmitry-viskov,

The upgrade of LTI got a bit pushed back at my work, and I had to focus on some other more pressing issues. Sorry about that, I will fix conflicts and rebase hopefully Monday or Tuesday at the latest. Thanks for your work and patience!

**Comment by dmitry-viskov** [comment]
hi @libre-man .

No rush! Feel free to update PR anytime when you will have free time.

Thanks!

**Comment by libre-man** [comment]
@dmitry-viskov I finally managed to finish this up! It should be all good to go.

File: pylti1p3/tool_config/abstract.py [hunk 6]
```diff
@@ -1,28 +1,54 @@
...
             return self.find_registration_by_issuer(iss)
 
     @abstractmethod
     def find_registration_by_issuer(self, iss, *args, **kwargs):
+        # type: (str, *None, **FIND_REG_KWARGS) -> t.Optional[Registration]
         raise NotImplementedError
 
     @abstractmethod
-    def find_deployment(self, iss, deployment_id):
+    def find_deployment(self, iss, deployment_id, get_param):
```
Comment (dmitry-viskov) [inline pylti1p3/tool_config/abstract.py]: `get_param` new param is related to my question in `message_launch.py`

File: pylti1p3/tool_config/dict.py [hunk 7]
```diff
@@ -26,7 +36,8 @@ def find_registration_by_issuer(self, iss, *args, **kwargs):
             .set_issuer(iss)\
             .set_tool_private_key(self.get_private_key(iss))
 
-    def find_deployment(self, iss, deployment_id):
+    def find_deployment(self, iss, deployment_id, get_param):  # pylint: disable=unused-argument
```
Comment (dmitry-viskov) [inline pylti1p3/tool_config/dict.py]: The same here

File: pylti1p3/message_launch.py [hunk 8]
```diff
@@ -345,17 +501,24 @@ def validate_jwt_signature(self):
         return self
 
     def validate_deployment(self):
+        # type: (T_SELF) -> T_SELF
         iss = self._get_iss()
         deployment_id = self._get_deployment_id()
 
         # Find deployment.
-        deployment = self._tool_config.find_deployment(iss, deployment_id)
```
[the remark this answers is not in the document]
Comment (libre-man) [reply pylti1p3/message_launch.py]: Oh I'm sorry this is an old artifact of the changes I did before, I don't think they are necessary but I will check tomorrow. If they are I will move them to another PR.

File: pylti1p3/grade.py [hunk 9]
```diff
@@ -59,4 +88,7 @@ def get_value(self):
             'timestamp': self._timestamp,
             'userId': self._user_id
         }
+        if self._extra_claims is not None:
```
[the remark this answers is not in the document]
Comment (libre-man) [reply pylti1p3/grade.py:91]: We tried to use it, but the LMS support really isn't there yet :(

**Comment by libre-man** [comment]
Hi @dmitry-viskov,

Sorry for the long delay, with the current crisis LTI1.3 support unfortunately moved a bit further back on our backlog :(.

I've updated the pull request for the new changes (awesome work btw), although I didn't find the time yet to really precisely type everything. I was wondering about a few changes though:
- It seems like the `set_request` method on the `Request` class was removed, but it is still relied upon (it seems) in some contrib classes. Should I add it back to the `Request` class (and make the class generic in the parameter to `set_request`).
- I don't really understand the `launch_data_storage` module, it seems like it is a generic key value storage for caching some values? But I don't really understand the way it is used in `oidc_login`, so I've made it generic in `int` there, but that probably is not correct.

**Comment by dmitry-viskov** [comment]
hi @libre-man.

* From the version 1.5.0 I've added ability to set more than one client id per one issuer in the config. You are right - it is useful in case of using cloud LMS-es where `iss` can't be changed.
 
* `set_request` was removed from the `Request` abstract class because it is only `Django`-specific case. For example in Flask `request` is global object so there is no need to implement this method in any custom `Request` class.

* the goal of the `launch_data_storage` is provide mechanism to save params like `nonce`, `state` not only using built-in django/flask framework session. The problem in framework's sessions is that there are too much cases when session-id could be set with params that can't be used if you want to embed you web page using iframe. For example - ability to set `SameSite: None` for session ID in Django will be available only from version 3.1: https://code.djangoproject.com/ticket/30862 . So it is much easier to manipulate with data using own `session-id` cookie and save data in cache. But anyway it is just an option. I mean that by default framework session will be used (i.e `SessionDataStorage`). Technically `LaunchDataStorage` class is some kind of key/value storage.

> I've made it generic in `int` there, but that probably is not correct

Sorry, not quite sure that understand what do you mean

**Comment by libre-man** [comment]
Thanks for the extensive update, that was really helpful!

So I've made the `launch_data_storage` generic, so that you can force it to only support specific value types (only in the type system, it doesn't introduce any runtime checks). However, as it isn't really clear to me what kind of types are stored in the store, for now I've made it possible to store any value in it, and retrieving something also returns the `Any` type. The large disadvantage of this is of course that you effectively lose your type safety, as you can do anything on a `Any` type, but I think for now that isn't such a big deal.

In conclusion, I think this PR is (finally) in a final state and ready to merge. So I'm looking forward to your review. And again thanks for your time, and patience :)

**Comment by dmitry-viskov** [comment]
hi @libre-man . Thanks for update and sorry for slow response. i'll try to find time to review all this in the nearest future.

**Comment by dmitry-viskov** [comment]
hi @libre-man . Sorry for delay. Finally I've merged your PR.
Thanks so much for the contribution! 👍
